core/src/main/java/org/elasticsearch/common/rounding/TimeZoneRounding.java

```diff
@@ -142,10 +142,20 @@ public long valueForKey(long time) {
 
         @Override
         public long nextRoundingValue(long time) {
-            long timeLocal = time;
-            timeLocal = timeZone.convertUTCToLocal(time);
-            long nextInLocalTime = durationField.add(timeLocal, 1);
-            return timeZone.convertLocalToUTC(nextInLocalTime, false);
+            if (unitIsDayOrLonger()) {
+                time = timeZone.convertUTCToLocal(time);
+            }
+            long next = durationField.add(time, 1);
+            if (unitIsDayOrLonger()) {
+                return timeZone.convertLocalToUTC(next, false);
+            } else {
+                return next;
+            }
+        }
+
+        private boolean unitIsDayOrLonger() {
+            return (unit == DateTimeUnit.HOUR_OF_DAY || unit == DateTimeUnit.MINUTES_OF_HOUR
+                    || unit == DateTimeUnit.SECOND_OF_MINUTE) == false;
```

Comment:
maybe we could do something like `unit.field().add(0, 1) >= number_of_millis_in_a_day`?